yksgi/src/lib.rs

```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+                    if r.id == smid {
+                        pinfo = Some(&entry.pinfo);
+                        rec = Some(r);
+                        break;
+                    }
+                }
+            };
+            let rec = rec.unwrap();
+            let pinfo = pinfo.unwrap();
+            if i > 0 {
+                // We don't need to include the bottom-most frame, which is the frame containing
+                // the control point. This frame still exists and thus doesn't need to be
+                // reconstructed.
+                memsize += usize::try_from(rec.size).unwrap();
+            }
+            memsize += 8; // Reserve space for return address.
+            smaps.push((i, frame, rec, pinfo));
+        }
+
+        // Add space to store the size of the stack which we'll need later to memcpy the correct
+        // amount.
+        memsize += 8;
+
+        // Now that we've calculated the stack's size, allocate it using mmap.
+        let mmap = unsafe {
+            libc::mmap(std::ptr::null_mut(), memsize, libc::PROT_READ | libc::PROT_WRITE, libc::MAP_PRIVATE | libc::MAP_ANONYMOUS, -1, 0)
+        };
+
+        // Get the modules layout which we'll need to extract type sizes of LLVM IR.
+        let layout = self.module.datalayout();
+
+        // Generate and write frames to the mmap. Since the stack grows downwards and we need to
+        // keep track of spilled register values we write to the mmap from back to front. To make
+        // things easier to think about we create two variables rbp and rsp which simulate their
+        // assembler namesakes.
+        // Note that while the bottom-most frame still exists and doesn't need to be reconstructed,
+        // we still need to get the register values from its stackmap in case those registers are
+        // spilled by the next frame.
+
+        let mut rbp = unsafe { mmap.offset(isize::try_from(memsize).unwrap()) };
+        let mut rsp = rbp;
+        // Keep track of the addresses of the current and previous frame so we can update the RBP
+        // register as needed.
+        let mut currframe = btmframeaddr;
+        let mut nextframe = btmframeaddr;
+
+        for (i, frame, rec, pinfo) in smaps {
+            debug_assert!(rec.size != u64::MAX);
+
+            // WRITE RBP
+            // If the current frame has pushed RBP we need to do the same (unless we are processing
+            // the bottom-most frame).
+            if pinfo.hasfp && i > 0 {
+                rsp = unsafe { rsp.offset(-8) };
+                rbp = rsp;
+                unsafe { ptr::write(rsp as *mut u64, currframe as u64) };
+            }
+
+            // Now that we've (potentially) written the last frame's address, update currframe to
+            // the actual current frame.
+            currframe = nextframe;
+
+            // Update RBP to represent this frame's address.
+            if pinfo.hasfp {
+                registers[6] = currframe as u64;
```

Comment:
Is this cast always safe?

Reply:
On x86_64 yes. Let me see if I can turn them into usizes.

Reply:
I could probably make `registers` store `usize`, but we are storing `SGValue.val` in them, which is `u64`. We had a discussion some time ago that `SGValue` probably shouldn't store `u64`s. But I don't think that's a change for this PR, and I'm also unsure how high on the list this should go.

Reply:
I was thinking the other way around. I understand that on x86_64 the registers can be treated as `u64`: I just wasn't entirely sure if `currframe` could be case to a `u64` in all cases or not. [I'm not sure what type `currframe` has. Knowing that might make it obvious.]

Reply:
Does this explanation resolve this for you?